File: 013 括号的匹配/main.swift
import Foundation

struct Stack<E> {
    var top: Int
    var maxsize: Int
    var elems: [E?]
    
    init(_ maxsize: Int = 20) {
        self.maxsize = maxsize
        self.elems = Array<E?>(repeating: nil, count: maxsize)
        self.top = 0
    }
    
    mutating func clear() {
        top = 0
    }
    
    func isEmpty() -> Bool {
        return top == 0
    }
    
    func length() -> Int {
        return top
    }
    
    func getTop() -> E? {
        return top > 0 ? elems[top - 1] : nil
    }
    
    mutating func push(_ e: E) -> Bool {
        if top >= maxsize {
            return false
        }
        elems[top] = e
        top += 1
        return true
    }
    
    mutating func pop() -> E? {
        if top > 0 {
            let e = elems[top - 1]
            top -= 1
            return e
        }
        return nil
    }
    
    func traverse() {
        for i in 0..<top {
            print(elems[i]!, terminator: " ")
        }
        print()
    }
}

func match(_ string: String) -> Bool {
    let map: [Character: Character] = [")": "(", "]": "[", "}": "{"]
    
    var stack = Stack<Character>()
    
    for c in string {
        switch c {
        case "(", "[", "{":
            _ = stack.push(c)
        case ")", "]", "}":
            if stack.getTop() == map[c] {
                _ = stack.pop()
            } else {
                return false
            }
        default:
            break
        }
    }
    return stack.isEmpty()
}

print(match("[((({))})]"))

File: 013 括号的匹配/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int size;
    int *base;
    int top;
} Stack;

Stack * createStack(int size) {
    Stack *stack = (Stack *)malloc(sizeof(Stack));
    stack->size = size;
    stack->base = malloc(sizeof(int) * size);
    stack->top = 0;
    return stack;
}

int getTop(Stack *stack) {
    if (stack->top == 0) {
        return -1000;
    }
    int idx = stack->top - 1;
    return stack->base[idx];
}

void push(Stack *stack, int e) {
    if (stack->top == stack->size) {
        return;
    }
    stack->base[stack->top] = e;
    stack->top += 1;
}

int pop(Stack *stack) {
    if (stack->top == 0) {
        return -10000;
    }
    stack->top -= 1;
    return stack->base[stack->top];
}

int isEmpty(Stack *stack) {
    return stack->top == 0;
}

int match(char *string) {
    Stack *stack = createStack(20);
    
    int i = 0;
    char c = string[i];
    
    while (c != '\n') {
        if (c == '(' || c == '[' || c == '{') {
            push(stack, c);
        } else if (c == ')') {
            if (!(getTop(stack) == '(')) {
                return 0;
            }
            pop(stack);
        } else if (c == ']') {
            if (!(getTop(stack) == '[')) {
                return 0;
            }
            pop(stack);
       } else if (c == '}') {
            if (!(getTop(stack) == '{')) {
                return 0;
            }
           pop(stack);
       }
        
        
        i++;
        c = string[i];
    }
    
    return isEmpty(stack);
}


int main() {
    
    char *test = "{dfasfa(32432fdasf])32324}\n";
    
    printf("%d \n", match(test));
    
    return 0;
}



